tools/src/org.graalvm.tools.lsp/src/org/graalvm/tools/lsp/server/TruffleAdapter.java

```diff
@@ -0,0 +1,421 @@
+        surrogate.setEditorText(text);
+        return surrogate;
+    }
+
+    public void didClose(URI uri) {
+        surrogateMap.remove(uri);
+    }
+
+    public Future<CallTarget> parse(final String text, final String langId, final URI uri) {
+        return contextAwareExecutor.executeWithDefaultContext(() -> parseWithEnteredContext(text, langId, uri));
+    }
+
+    protected CallTarget parseWithEnteredContext(final String text, final String langId, final URI uri) throws DiagnosticsNotification {
+        LanguageInfo languageInfo = findLanguageInfo(langId, uri);
+        TextDocumentSurrogate surrogate = getOrCreateSurrogate(uri, text, languageInfo);
+        return parseWithEnteredContext(surrogate);
+    }
+
+    CallTarget parseWithEnteredContext(TextDocumentSurrogate surrogate) throws DiagnosticsNotification {
+        return sourceCodeEvaluator.parse(surrogate);
+    }
+
+    public Future<?> reparse(URI uri) {
+        TextDocumentSurrogate surrogate = surrogateMap.get(uri);
+        return contextAwareExecutor.executeWithDefaultContext(() -> parseWithEnteredContext(surrogate));
+    }
+
+    /**
+     * Special handling needed, because some LSP clients send a MIME type as langId.
+     *
+     * @param langId an id for a language, e.g. "sl" or "python", or a MIME type
+     * @param uri of the concerning file
+     * @return a language info
+     */
+    private LanguageInfo findLanguageInfo(final String langId, final URI uri) {
+        if (env.getLanguages().containsKey(langId)) {
+            return env.getLanguages().get(langId);
+        }
+
+        String possibleMimeType = langId;
+        String actualLangId = Source.findLanguage(possibleMimeType);
+        if (actualLangId == null) {
+            try {
+                actualLangId = Source.findLanguage(uri.toURL());
+            } catch (IOException e) {
+            }
+
+            if (actualLangId == null) {
+                actualLangId = langId;
+            }
+        }
+
+        if (!env.getLanguages().containsKey(actualLangId)) {
+            throw new UnknownLanguageException("Unknown language: " + actualLangId + ". Known languages are: " + env.getLanguages().keySet());
+        }
+
+        return env.getLanguages().get(actualLangId);
+    }
+
+    public Future<TextDocumentSurrogate> processChangesAndParse(List<? extends TextDocumentContentChangeEvent> list, URI uri) {
+        return contextAwareExecutor.executeWithDefaultContext(() -> processChangesAndParseWithContextEntered(list, uri));
+    }
+
+    protected TextDocumentSurrogate processChangesAndParseWithContextEntered(List<? extends TextDocumentContentChangeEvent> list, URI uri) throws DiagnosticsNotification {
+        TextDocumentSurrogate surrogate = surrogateMap.get(uri);
+
+        if (surrogate == null) {
+            throw new IllegalStateException("No internal mapping for uri=" + uri.toString() + " found.");
```

Comment:
Shouldn't it load and parse the file instead? When I create a new file in VS Code I get this exception.

Reply:
Yes, already fixed.